config.org
#+title:GNU EMACS CONFIGURATION
#+author:Shawn Martin-Truesdell
#+email:shawn@martin-truesdell.com

This file is written in [[http://www.howardism.org/Technical/Emacs/literate-programming-tutorial.html][literate programming style]] using [[https://orgmode.org/][org-mode]].

See =./config.el= for the generated file.

* Emacs Initialization

** Preamble

#+begin_src emacs-lisp :tangle yes
  ;;; mt-config --- Shawn Martin-Truesdell's Emacs Configuration
  ;;; License:
  ;; ---------------------------------------------------------------------
  ;; Copyright (C) 2022 - Shawn Martin-Truesdell
  ;;
  ;; This program is free software; you can redistribute it and/or modify
  ;; it under the terms of the GNU General Public License as published by
  ;; the Free Software Foundation, either version 3 of the License, or
  ;; (at your option) any later version.
  ;;
  ;; This program is distributed in the hope that it will be useful,
  ;; but WITHOUT ANY WARRANTY; without even the implied warranty of
  ;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  ;; GNU General Public License for more details.
  ;;
  ;; You should have received a copy of the GNU General Public License
  ;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
  ;;
  ;; ---------------------------------------------------------------------
  ;;; Commentary:
  ;; DO NOT EDIT THIS FILE DIRECTLY
  ;; This is a file generated from a literate programing source file.
  ;; You should make any changes there and regenerate it from Emacs org-mode
  ;; using C-c C-v t
  ;;
  ;; ---------------------------------------------------------------------
  ;;; Code:
#+end_src

** Straight.el

=straight.el= is a replacement for =package.el=, which is the default package manager for EMACs. It's pretty fancy, and instead of pulling compiled blobs it synchronizes local repositories for each package and allows for easy local editing of those repositories.

#+begin_src emacs-lisp :tangle yes
  (require 'package)                   ; Bring in to the environment all package management functions

  ;; A list of package repositories
  (setq package-archives '(("melpa" . "https://melpa.org/packages/")
			   ("org"   . "https://orgmode.org/elpa/")
			   ("elpa"  . "https://elpa.gnu.org/packages/")))

  (unless package-archive-contents     ; Unless a package archive already exists,
    (package-refresh-contents))        ; Refresh package contents so that Emacs knows which packages to load


  ;; Initialize use-package on non-linux platforms
  (unless (package-installed-p 'use-package)        ; Unless "use-package" is installed, install "use-package"
    (package-install 'use-package))

  (require 'use-package)                            ; Once it's installed, we load it using require

  ;; Make sure packages are downloaded and installed before they are run
  ;; also frees you from having to put :ensure t after installing EVERY PACKAGE.
  (setq use-package-always-ensure t)

  (defvar bootstrap-version)
  (let ((bootstrap-file
	 (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
	(bootstrap-version 5))
    (unless (file-exists-p bootstrap-file)
      (with-current-buffer
	  (url-retrieve-synchronously
	   "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
	   'silent 'inhibit-cookies)
	(goto-char (point-max))
	(eval-print-last-sexp)))
    (load bootstrap-file nil 'nomessage))
#+end_src

** use-package

To be able to manage its configuration with =use-package= it is necessary to
first install it

#+BEGIN_SRC emacs-lisp :tangle yes
  (straight-use-package 'use-package)
  (setq straight-use-package-by-default t)
  (setq use-package-always-demand t)
  (package-initialize)
  (use-package dash)
#+END_SRC

* Start Up
** exec-path-from-shell

We don't want Emacs to use a different PATH setting from the rest of our
OS, so we'll take advantage of the =exec-path-from-shell= package.

#+BEGIN_SRC emacs-lisp :tangle yes
  (use-package exec-path-from-shell
    :if (memq window-system '(mac ns))
    :config
    (setq exec-path-from-shell-variables '("PATH"))
    (exec-path-from-shell-initialize))
#+END_SRC

** Better Defaults

For a better user experience of GNU Emacs, here are the default values I use.

#+BEGIN_SRC emacs-lisp :tangle yes
  (setq-default
   ad-redefinition-action 'accept                   ; Silence warnings for redefinition
   cursor-in-non-selected-windows t                 ; Hide the cursor in inactive windows
   display-time-default-load-average nil            ; Don't display load average
   fill-column 80                                   ; Set width for automatic line breaks
   help-window-select t                             ; Focus new help windows when opened
   indent-tabs-mode nil                             ; Use spaces instead of tabs
   inhibit-startup-screen t                         ; Disable start-up screen
   initial-scratch-message ""                       ; Empty the initial *scratch* buffer
   kill-ring-max 128                                ; Maximum length of kill ring
   load-prefer-newer t                              ; Prefers the newest version of a file
   mark-ring-max 128                                ; Maximum length of mark ring
   scroll-conservatively most-positive-fixnum       ; Always scroll by one line
   select-enable-clipboard t                        ; Merge system's and Emacs' clipboard
   tab-width 2                                      ; Set width for tabs
   user-full-name "Shawn Martin-Truesdell"          ; Set the full name of the current user
   user-mail-address "shawn@martin-truesdell.com"   ; Set the email address of the current user
   vc-follow-symlinks t                             ; Always follow the symlinks
   view-read-only t)                                ; Always open read-only buffers in view-mode
  (cd "~/")                                         ; Move to the user directory
  (column-number-mode 1)                            ; Show the column number
  (fset 'yes-or-no-p 'y-or-n-p)                     ; Replace yes/no prompts with y/n
  (global-hl-line-mode)                             ; Hightlight current line
  (set-default-coding-systems 'utf-8)               ; Default to utf-8 encoding
  (show-paren-mode 1)                               ; Show the parent
  (setq backup-directory-alist '(("." . "~/.emacs-saves")))
  (setq ring-bell-function 'ignore)                 ; No more dings
  (global-unset-key (kbd "C-z"))                    ; No more minimizing on accident
  ;(pixel-scroll-precision-mode) ;; emacs 29
#+END_SRC

** Better file management with XDG

To keep the user's home and the =~/.emacs.d= folder as clean as possible, I
follow the [[https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html][XDG base directory specification]].

Be careful that GNU Emacs will not create the appropriate folders if they do not
exist. Therefore, it is necessary to create them yourself:

#+BEGIN_SRC bash
  mkdir ~/.cache/emacs
#+END_SRC

#+BEGIN_SRC emacs-lisp :tangle yes
  (defun default-to (fb val)
    (if (eq nil val)
        fb
      val))
  (defvar xdg-bin (default-to "~/.local/bin" (getenv "XDG_BIN_HOME"))
    "The XDG bin base directory.")

  (defvar xdg-cache (default-to "~/.cache" (getenv "XDG_CACHE_HOME"))
    "The XDG cache base directory.")

  (defvar xdg-config (default-to "~/.config" (getenv "XDG_CONFIG_HOME"))
    "The XDG config base directory.")

  (defvar xdg-data (default-to "~/.local/share" (getenv "XDG_DATA_HOME"))
    "The XDG data base directory.")

  (defvar xdg-lib (default-to "~/.local/lib" (getenv "XDG_LIB_HOME"))
    "The XDG lib base directory.")
#+END_SRC

** Customization

To avoid overloading the GNU Emacs custormization =init.el= file made by the
user with the UI, I add the generated code in a separate file.

I also made sure to follow the XDG base directory specification for the
=auto-save-file= folder, in order to keep my =~/.emacs.d= folder clean.

#+BEGIN_SRC emacs-lisp :tangle yes
  (setq-default
   ;; Autosave Directory
   auto-save-list-file-name (expand-file-name (format "%s/emacs/auto-save-list" xdg-data))
   ;; Session customizations
   custom-file (expand-file-name (format "%s/emacs/custom.el" xdg-data)))
  (when (file-exists-p custom-file)
    (load custom-file t))
#+END_SRC

** Keep Autosave/Temp files corraled

Emacs fills your directories with =#filename= temporary files and =filename~= backups, and I hate it.
Instead, I want those to be kept together in a separate folder so that my source directories remain uncluttered.

#+begin_src emacs-lisp :tangle yes
  ;;; my/set-autosave-file-directory --- Sets the autosave/temp file directory
  ;;; Commentary:
  ;;; Code:
  (defun my/set-autosave-file-directory ()
    "Set the autosave/temp file directory."
    (let ((user-temporary-file-directory (concat temporary-file-directory user-login-name "/")))
      (make-directory user-temporary-file-directory t)
      (setq create-lockfiles nil)
      ;; Lockfiles are creating the '.#______' files that are messing up dev servers
      (setq backup-by-copying t)
      (setq backup-directory-alist
            `(("." . ,user-temporary-file-directory)
              (,tramp-file-name-regexp nil)))
      (setq auto-save-list-file-prefix
            (concat user-temporary-file-directory ".auto-saves-"))
      (setq auto-save-file-name-transforms
            `((".*" ,(expand-file-name "\\2" user-temporary-file-directory) t)))))

  (my/set-autosave-file-directory)
#+end_src


** No-Alt Extended Commands

=M-x= is a chord we use ALL THE TIME, and having it right under the center of our left palm is super bad for our ergonomics. Instead, we're going to bind a new chord to make it more convenient.

#+BEGIN_SRC emacs-lisp :tangle yes
  (global-set-key "\C-x\C-m" 'execute-extended-command)
  (global-set-key "\C-c\C-m" 'execute-extended-command)
#+END_SRC

** Prefer backward-kill-word

#+begin_quote
Emacs Wizards try to avoid hitting the backspace key, because it's just remote enough from home-row to be annoying. We make typing mistakes all the time, but if you type faster than about 50 wpm, it's more economical to kill the entire word and re-type it than to painstakingly backspace to your mistake
#+end_quote

#+BEGIN_SRC emacs-lisp :tangle yes
  (global-set-key "\C-w" 'backward-kill-word)
  (global-set-key "\C-x\C-k" 'kill-region)
  (global-set-key "\C-c\C-k" 'kill-region)
#+END_SRC

** Fix the Frame Title

Such a silly little thing, but I can't get rid of the frame title bar so I'm going to at least make it a little nicer.

#+BEGIN_SRC emacs-lisp :tangle yes
  (setq-default frame-title-format '("Shawn's Emacs :: %b"))
#+END_SRC

** Theme

#+BEGIN_SRC emacs-lisp :tangle yes
  (use-package nano-theme
    :straight (nano-theme :type git :host github :repo "rougier/nano-theme")
    :custom
    (nano-fonts-use t)
    (nano-light-salient "SteelBlue3")
    :init (load-theme 'nano-light t))

  (use-package nano-modeline
    :straight (nano-modeline :type git :host github :repo "rougier/nano-modeline")
    :init (nano-modeline-mode))
#+end_src

** Turn off mouse interface

Since I never use the mouse with GNU Emacs, I prefer not to use certain
graphical elements as seen as the menu bar, toolbar, scrollbar and toolt
I find invasive.

#+BEGIN_SRC emacs-lisp :tangle yes
  (when window-system
    (menu-bar-mode -1)              ; Disable the menu bar
    (scroll-bar-mode -1)            ; Disable the scroll bar
    (tool-bar-mode -1)              ; Disable the tool bar
    (tooltip-mode -1))              ; Disable the tooltips
#+END_SRC


* Org Mode

Org-Mode is the killer feature, and after basic text editing ergonomics it is THE main reason to use Emacs. We love it for its organizational tools, it's presentational tools, and for the awesome power of literate programming.

#+BEGIN_SRC emacs-lisp :tangle yes :noweb no-export
  (use-package org-contrib)
  (use-package org
      :after org-contrib
      :init
      <<org-mode-fn-definitions>>
      :bind
      (:map org-mode-map
            <<org-mode-keybindings>>)
      :custom-face
      <<org-mode-faces>>
      :hook
      <<org-mode-hooks>>
      :custom
      <<org-mode-custom-variables>>
      :config
      <<org-mode-config>>)

    ;; Maintains a clean TOC in the first section with the :TOC: tag.
    (use-package toc-org
      :after org
      :hook (org-mode . toc-org-enable))
    ;; Cleans up "online mode" when connecting using TRAMP
    (use-package org-indent :straight nil :ensure nil :after org)
#+END_SRC

** Utility Function Definitions

*** Compare Logged and Estimated times

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-fn-definitions
  (defun my/org-compare-times (clocked estimated)
    "Gets the ratio between the timed time and the estimated time."
    (if (and (> (length clocked) 0) estimated)
        (format "%.2f"
                (/ (* 1.0 (org-hh:mm-string-to-minutes clocked))
                   (org-hh:mm-string-to-minutes estimated)))
      ""))
#+END_SRC

*** Automagically archive finished tasks

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-fn-definitions
  (defun my/org-archive-done-tasks ()
    "Archives finished or cancelled tasks."
    (interactive)
    (org-map-entries
     (lambda ()
       (org-archive-subtree)
       (setq org-map-continue-from (outline-previous-heading)))
     "TODO=\"DONE\"|TODO=\"CANCELLED\"" (if (org-before-first-heading-p) 'file 'tree)))
#+END_SRC

** Hooks

*** Auto-Tangle when saving the config file
#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-hooks
  (after-save . my/config-tangle)
#+END_SRC

*** Auto-Indent when loading an Org file

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-hooks
  (org-mode . org-indent-mode)
#+END_SRC

** Boilerplate Customizations

*** Set file paths

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-archive-location "~/.personal/archives/%s::")
  (org-ditaa-jar-path "~/.local/lib/ditaa0_9.jar")
#+END_SRC

*** Load Org modules to add new features

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-modules '(org-habit
                 org-info
                 org-mouse
                 org-protocol))
#+END_SRC

*** Tags and Todo Settings

TODO: I should review these, as I basically never use them anymore. Maybe I will in a laptop world

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-tag-alist '(("@work" . ?w)
                   ("@errands" . ?e)
                   ("@home" . ?h)
                   ("@kiddo" . ?k)
                   ("@phone" . ?p)
                   ("@reading" . ?r)
                   ("@admin" . ?a)
                   ("fuzzy" . ?0)))
  (org-tags-exclude-from-inheritance '("crypt" "project"))
  (org-todo-keywords '((sequence "TODO(t)"
                                 "STARTED(s)"
                                 "WAITING(w@/!)"
                                 "SOMEDAY(.)" "|" "DONE(x!)" "CANCELLED(c@)")))
#+END_SRC

*** Set up Org Refile to make it a little more robust

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-refile-allow-creating-parent-nodes 'confirm)
  (org-refile-use-cache nil)
  (org-refile-use-outline-path t)
  (org-refile-targets '((org-agenda-files . (:maxlevel . 6))))
#+END_SRC

*** Automatically log done times

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-log-done 'time)
  (org-log-into-drawer "LOGBOOK")
#+END_SRC

*** Open src blocks in the same window

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-src-window-setup 'current-window)
#+END_SRC

*** Hide Emphasis (bold, fixed-width, etc) Markers

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-hide-emphasis-markers t)
#+END_SRC

*** Other settings

#+BEGIN_SRC emacs-lisp :tangle no :noweb-ref org-mode-custom-variables
  (org-blank-before-new-entry '((heading . t)))
  (org-cycle-include-plain-lists 'integrate)
  (org-expiry-inactive-timestamps t)
  (org-export-backends '(ascii html latex md))
  (org-startup-folded nil)
  (org-startup-indented t)
  (org-startup-with-inline-images t)
  (org-use-effective-time t)
  (org-yank-adjusted-subtrees t)
#+END_SRC

** Beautification
*** Enable mixed pitch to allow src blocks in fixed-pitch

This elisp function lets us set the face without clearing out other customizations (colors, underscores, etc).

TODO: Check if this is still necessary.

#+begin_src emacs-lisp :tangle no :noweb-ref org-mode-config
  (defun my-adjoin-to-list-or-symbol (element list-or-symbol)

config.el
;; Package Management - Straight.el
(setq straight-check-for-modifications nil)

(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

;; (straight-pull-recipe-repositories) ;; only necessary if recipes are missing.

;; Packages installation

(setq package-list
      '(cape                ; Completion At Point Extensions
        orderless           ; Completion style for matching regexps in any order
        vertico             ; VERTical Interactive COmpletion
        marginalia          ; Enrich existing commands with completion annotations
        avy                 ; Cursor/Pointer interaction FSA library
        consult             ; Consulting completing-read
        corfu               ; Completion Overlay Region FUnction
        deadgrep            ; superfast grep
        bind-key            ; easy keybindings
        move-text           ; easy up/down moving
        ;; deft                ; Quickly browse, filter, and edit plain text notes
        f                   ; Modern API for working with files and directories
        ;; citar               ; Citation-related commands for org, latex, markdown
        ;; citeproc            ; A CSL 1.0.2 Citation Processor
        flyspell-correct-popup ; Correcting words with flyspell via popup interface
        flyspell-popup      ; Correcting words with Flyspell in popup menus
        guess-language      ; Robust automatic language detection
        helpful             ; A better help buffer
        htmlize             ; Convert buffer text and decorations to HTML
        mini-frame          ; Show minibuffer in child frame on read-from-minibuffer
        olivetti            ; Smart margins for better writing
        ;; imenu-list          ; Show imenu entries in a separate buffer
        magit               ; A Git porcelain inside Emacs.
	      forge               ; Forge (GitHub, GitLab, etc) integrations for Magit
        git-gutter          ; change markers in the gutter
        markdown-mode       ; Major mode for Markdown-formatted text
        ;; multi-term          ; Managing multiple terminal buffers in Emacs.D
        ;; pinentry            ; GnuPG Pinentry server implementation
        use-package         ; A configuration macro for simplifying your .emacs
        vc-backup           ; VC backend for versioned backups
        yaml-mode           ; YAML mode
        consult-recoll      ; Consult interface for recoll query
        ;; org-auto-tangle     ; Tangle org file when it is saved
        exec-path-from-shell; Get environment variables such as $PATH from the shell
        nano-modeline       ; NANO customized modeline
        tabnine             ; Awesome AI powered autocompletion
        typescript-mode     ; TypeScript
        web-mode            ; JSX-enabled web mode
        apheleia            ; Format-on-Save for code buffers
        eglot               ; language-server backing for Emacs
        yasnippet           ; Awesome snippets
        yasnippet-snippets  ; Presets
        cargo               ; Rust cargo minor-mode
        smartparens         ; Only somewhat smart
        hungry-delete       ; eat all the whitespace
        switch-window       ; smart window switching
        all-the-icons       ; cute icons
        all-the-icons-dired ; cute icons, dired edition
        which-key))         ; Display available keybindings in popup

;; Install packages that are not yet installed
(dolist (package package-list)
  (straight-use-package package))

(require 'exec-path-from-shell)
(setq exec-path-from-shell-variables '("PATH"))
(exec-path-from-shell-initialize)

;; Special case for pdf-tools that has recently (2022) changed maintainer
(straight-use-package
 '(pdf-tools :type git :host github :repo "vedang/pdf-tools"))

(straight-use-package
 '(tsi :type git :host github :repo "orzechowskid/tsi.el"))

(require 'bind-key)

;; TabNine

;; === BEGIN TABNINE KEY ===
;; shawntruesdell@gmail.com
;; fpctaeuujatzulocpebuumbdfkbbfqaj
;; === END TABNINE KEY ===

(use-package tabnine
  :commands (tabnine-start-process)
  :hook (prog-mode . tabnine-mode)
  :diminish "⌬"
  :custom
  (tabnine-wait 1)
  (tabnine-minimum-prefix-length 0)
  :hook (kill-emacs . tabnine-kill-process)
  :config

  (tabnine-start-process)
  :bind
  (:map tabnine-completion-map
	      ("<tab>" . tabnine-accept-completion)
	      ("TAB" . tabnine-accept-completion)
	      ("M-f" . tabnine-accept-completion-by-word)
	      ("M-<return>" . tabnine-accept-completion-by-line)
	      ("C-g" . tabnine-clear-overlay)
	      ("M-[" . tabnine-previous-completion)
	      ("M-]" . tabnine-next-completion)))

;; NANO modules

;; - NANO splash
(straight-use-package
 '(nano-splash :type git :host github :repo "rougier/nano-splash"))

(require 'nano-splash)

(straight-use-package
 '(nano-theme :type git :host github :repo "rougier/nano-theme"))

(require 'nano-theme)
(setq nano-fonts-use t) ; Use theme font stack
(setq nano-light-salient "SteelBlue3")
(nano-light)

(set-face-attribute 'default nil
                    :family "Roboto Mono"
                    :weight 'light
                    :height 140)

(set-face-attribute 'bold nil
                    :family "Roboto Mono"
                    :weight 'regular)

;; TODO: There's an extra step to hack the font to fit. :(
;; (set-face-attribute 'italic nil
;;                     :family "Victor Mono"
;;                     :weight 'semilight
;;                     :slant 'italic)

(set-fontset-font t 'unicode
                  (font-spec :name "Inconsolata Light"
                             :size 16) nil)

(set-fontset-font t '(#xe000 . #xffdd)
                  (font-spec :name "RobotoMono Nerd Font"
                             :size 12) nil)

;; Completion
(require 'corfu)

(setq corfu-cycle t                ; Enable cycling for `corfu-next/previous'
      corfu-auto t                 ; Enable auto completion
      corfu-auto-delay 60.0        ; Delay before auto-completion shows up
      corfu-separator ?\s          ; Orderless field separator
      corfu-quit-at-boundary nil   ; Never quit at completion boundary
      corfu-quit-no-match t        ; Quit when no match
      corfu-preview-current nil    ; Disable current candidate preview
      corfu-preselect-first nil    ; Disable candidate preselection
      corfu-on-exact-match nil     ; Configure handling of exact matches
      corfu-echo-documentation nil ; Disable documentation in the echo area
      corfu-scroll-margin 5)       ; Use scroll margin

(global-corfu-mode)

;; TAB cycle if there are only few candidates
(setq completion-cycle-threshold 3)

;; Emacs 28: Hide commands in M-x which do not apply to the current mode.
;; Corfu commands are hidden, since they are not supposed to be used via M-x.
(setq read-extended-command-predicate
      #'command-completion-default-include-p)

;; Enable indentation+completion using the TAB key.
;; completion-at-point is often bound to M-TAB.
(setq tab-always-indent 'complete)

;; Completion in source blocks
(require 'cape)

(add-to-list 'completion-at-point-functions 'cape-symbol)
(add-to-list 'completion-at-point-functions 'tabnine-completion-at-point)

(require 'orderless)

(setq completion-styles '(substring orderless basic)
      orderless-component-separator 'orderless-escapable-split-on-space
      read-file-name-completion-ignore-case t
      read-buffer-completion-ignore-case t
      completion-ignore-case t)

(require 'consult)

(setq consult-preview-key nil) ; No live preview

(bind-key "C-x C-r" #'consult-recent-file)
(bind-key "C-x h"   #'consult-outline)
(bind-key "C-x b"   #'consult-buffer)
(bind-key "C-c h"   #'consult-history)

(require 'vertico)

;; (setq completion-styles '(basic substring partial-completion flex))

(setq vertico-resize nil        ; How to resize the Vertico minibuffer window.
      vertico-count 8           ; Maximal number of candidates to show.
      vertico-count-format nil) ; No prefix with number of entries

(vertico-mode)

(require 'marginalia)

(setq-default marginalia--ellipsis "…"    ; Nicer ellipsis
              marginalia-align 'right     ; right alignment
              marginalia-align-offset -1) ; one space on the right

(marginalia-mode)

;; Startup

(setq-default
 inhibit-startup-screen t               ; Disable start-up screen
 inhibit-startup-message t              ; Disable startup message
 inhibit-startup-echo-area-message t    ; Disable initial echo message
 initial-scratch-message ""             ; Empty the initial *scratch* buffer
 initial-buffer-choice t)               ; Open *scratch* buffer at init

;; Encoding

(set-default-coding-systems 'utf-8)     ; Default to utf-8 encoding
(prefer-coding-system       'utf-8)     ; Add utf-8 at the front for automatic detection.
(set-terminal-coding-system 'utf-8)     ; Set coding system of terminal output
(set-keyboard-coding-system 'utf-8)     ; Set coding system for keyboard input on TERMINAL
(set-language-environment "English")    ; Set up multilingual environment

;; Recovery & Backups

(setq auto-save-list-file-prefix ; Prefix for generating auto-save-list-file-name
      (expand-file-name ".auto-save-list/.saves-" user-emacs-directory)
      auto-save-default t        ; Auto-save every buffer that visits a file
      auto-save-timeout 20       ; Number of seconds between auto-save
      auto-save-interval 200)    ; Number of keystrokes between auto-saves

(setq backup-directory-alist       ; File name patterns and backup directory names.
      `(("." . ,(expand-file-name "backups" user-emacs-directory)))
      make-backup-files t          ; Backup of a file the first time it is saved.
      vc-make-backup-files t       ; No backup of files under version contr
      backup-by-copying t          ; Don't clobber symlinks
      version-control t            ; Version numbers for backup files
      delete-old-versions t        ; Delete excess backup files silently
      kept-old-versions 6          ; Number of old versions to keep
      kept-new-versions 9          ; Number of new versions to keep
      delete-by-moving-to-trash t) ; Delete files to trash

;; - Back
(require 'vc-backup)

;; Bookmarks

(setq bookmark-default-file (expand-file-name "bookmark" user-emacs-directory))

;; Defaults

(setq-default
 ad-redefinition-action 'accept                   ; Silence warnings for redefinition
 cursor-in-non-selected-windows t                 ; Hide the cursor in inactive windows
 display-time-default-load-average nil            ; Don't display load average
 fill-column 80                                   ; Set width for automatic line breaks
 help-window-select t                             ; Focus new help windows when opened
 indent-tabs-mode nil                             ; Use spaces instead of tabs
 inhibit-startup-screen t                         ; Disable start-up screen
 initial-scratch-message ""                       ; Empty the initial *scratch* buffer
 kill-ring-max 128                                ; Maximum length of kill ring
 load-prefer-newer t                              ; Prefers the newest version of a file
 mark-ring-max 128                                ; Maximum length of mark ring
 scroll-conservatively most-positive-fixnum       ; Always scroll by one line
 select-enable-clipboard t                        ; Merge system's and Emacs' clipboard
 tab-width 2                                      ; Set width for tabs
 user-full-name "Shawn Martin-Truesdell"          ; Set the full name of the current user
 user-mail-address "shawn@martin-truesdell.com"   ; Set the email address of the current user
 vc-follow-symlinks t                             ; Always follow the symlinks
 view-read-only t)                                ; Always open read-only buffers in view-mode

(cd "~/")                                         ; Move to the user directory

(column-number-mode 1)                            ; Show the column number
(fset 'yes-or-no-p 'y-or-n-p)                     ; Replace yes/no prompts with y/n
(global-hl-line-mode)                             ; Hightlight current line
(set-default-coding-systems 'utf-8)               ; Default to utf-8 encoding
(show-paren-mode 1)                               ; Show the parent
(setq ring-bell-function 'ignore)                 ; No more dings
(global-unset-key (kbd "C-z"))                    ; No more minimizing on accident

;; History

(defun unpropertize-kill-ring ()
  (setq kill-ring (mapcar 'substring-no-properties kill-ring)))

(add-hook 'kill-emacs-hook 'unpropertize-kill-ring)

(require 'savehist)

(setq kill-ring-max 50
      history-length 50)

(setq savehist-additional-variables
      '(kill-ring
        command-history
        set-variable-value-history
        custom-variable-history
        query-replace-history
        read-expression-history
        minibuffer-history
        read-char-history
        face-name-history
        bookmark-history
        file-name-history))

(put 'minibuffer-history         'history-length 50)
(put 'file-name-history          'history-length 50)
(put 'set-variable-value-history 'history-length 25)
(put 'custom-variable-history    'history-length 25)
(put 'query-replace-history      'history-length 25)
(put 'read-expression-history    'history-length 25)
(put 'read-char-history          'history-length 25)
(put 'face-name-history          'history-length 25)
(put 'bookmark-history           'history-length 25)

(setq history-delete-duplicates t)

(let (message-log-max)
  (savehist-mode))

(setq save-place-file (expand-file-name "saveplace" user-emacs-directory)
      save-place-forget-unreadable-files t)

(save-place-mode 1)

;; Custom

(setq custom-file (concat user-emacs-directory "custom.el"))

(when (file-exists-p custom-file)
  (load custom-file nil t))

;; Fix meta =M-x= is a chord we use ALL THE TIME, and having it right
;; under the center of our left palm is super bad for our
;; ergonomics. Instead, we're going to bind a new chord to make it
;; more convenient.
(global-set-key "\C-x\C-m" 'execute-extended-command)
(global-set-key "\C-c\C-m" 'execute-extended-command)

;; OS Specific customizations
(cond
 ((string-equal system-type "windows-nt") ; Microsoft Windows
  (progn
    (message "Microsoft Windows")))
 ((string-equal system-type "darwin") ; Mac OS X
  (progn
    ;; Macbook Keyboard
    (setq mac-option-key-is-meta nil)
    (setq mac-command-key-is-meta t)
    (setq mac-command-modifier 'meta)
    (setq mac-option-modifier 'option)
    ;; Mechanical Keyboard
    ;; (setq mac-option-key-is-meta t)
    ;; (setq mac-command-key-is-meta nil)
    ;; (setq mac-command-modifier 'option)
    ;; (setq mac-option-modifier 'meta)
    (message "Mac OS X")))
 ((string-equal system-type "gnu/linux") ; linux
  (progn
    (message "Linux"))))

(when window-system
  (menu-bar-mode -1)              ; Disable the menu bar
  (scroll-bar-mode -1)            ; Disable the scroll bar
  (tool-bar-mode -1)              ; Disable the tool bar
  (tooltip-mode -1))              ; Disable the tooltips

;; Software Engineering

;; - Syntax Highlighting
(require 'treesit)

(global-tree-sitter-mode)
(add-to-list 'treesit-language-source-alist
             '(typescript "https://github.com/tree-sitter/tree-sitter-typescript" "v0.20.3" "typescript/src"))
(add-to-list 'treesit-language-source-alist
             '(tsx "https://github.com/tree-sitter/tree-sitter-typescript" "v0.20.3" "tsx/src"))
;; (setq treesit-load-name-override-list '((tsx "libtree-sitter-tsx" "tree_sitter_typescript")))

(when
    (not (treesit-language-available-p 'typescript))
  (treesit-install-language-grammar 'typescript))
(when
    (not (treesit-language-available-p 'tsx))
  (treesit-install-language-grammar 'tsx))

(require 'typescript-mode)
(add-to-list 'auto-mode-alist '("\\.tsx?" . tsx-ts-mode))
(add-to-list 'tree-sitter-major-mode-language-alist '(tsx-ts-mode . typescript))

(require 'tsi)
(add-hook 'tsx-ts-mode-hook (lambda () (tsi-typescript-mode 1)))
(add-hook 'typescript-mode-hook (lambda () (tsi-typescript-mode 1)))
(add-hook 'json-mode-hook (lambda () (tsi-json-mode 1)))
(add-hook 'css-mode-hook (lambda () (tsi-css-mode 1)))
(add-hook 'scss-mode-hook (lambda () (tsi-scss-mode 1)))

(require 'eglot)
(add-to-list 'eglot-server-programs
             '((typescript-mode) "typescript-language-server" "--stdio")
             '((tsx-ts-mode) "typescript-language-server" "--stdio"))
(add-to-list 'eglot-server-programs
             `(rust-ts-mode . ("rust-analyzer" :initializationOptions
                               ( :procMacro (:enable t)
                                 :cargo ( :buildScripts (:enable t)
                                          :features "all")))))

(bind-key "C-x ." 'eglot-code-actions)

(require 'apheleia)
(apheleia-global-mode +1)

(require 'deadgrep)
(keymap-global-set "<f5>" 'deadgrep)

(use-package dired
  :straight nil
  :ensure nil
  :bind (:map dired-mode-map ("M-+" . dired-create-empty-file))
  :custom
  (dired-auto-revert-buffer t)
  (dired-dwim-target t)
  (dired-hide-details-hide-symlink-targets nil)
  (dired-listing-switches "-alh")
  (dired-ls-F-marks-symlinks nil)
  (dired-recursive-copies 'always)
  :init
  (add-hook 'dired-mode-hook
            (lambda ()
              (dired-hide-details-mode))))

(use-package move-text
  :bind (("M-p" . move-text-up)
         ("M-n" . move-text-down))
  :config (move-text-default-bindings))


(require 'all-the-icons)
(use-package all-the-icons-dired
  :init
  (add-hook 'dired-mode-hook 'all-the-icons-dired-mode))

(use-package smartparens
  :custom (sp-escape-quotes-after-insert nil)
  :config (smartparens-global-mode 1))

(use-package hungry-delete :config (global-hungry-delete-mode))

(defun my/hsplit-last-buffer ()
  "Gives the focus to the last created horizontal window."
  (interactive)
  (split-window-horizontally)
  (other-window 1))
(defun my/vsplit-last-buffer ()
  "Gives the focus to the last created vertical window."
  (interactive)
  (split-window-vertically)
  (other-window 1))
(bind-key "C-x 3" 'my/hsplit-last-buffer)
(bind-key "C-x 2" 'my/vsplit-last-buffer)

(use-package switch-window
  :straight nil
  :bind (("C-x o" . switch-window)
         ("C-x w" . switch-window-then-swap-buffer)))

(use-package avy
  :bind (("M-j" . avy-goto-char-timer)))

(add-hook 'before-save-hook 'delete-trailing-whitespace)

(require 'magit)
(require 'forge)
(require 'git-gutter)
(global-git-gutter-mode +1)

(yas-global-mode)

(bind-key "C-w" 'backward-kill-word)
(bind-key "C-x C-k" 'kill-region)
(bind-key "C-c C-k" 'kill-region)

;; This causes Emacs to crash when restoring from the background, somehow
;;(setq-default frame-title-format '("Shawn's Emacs :: %b"))

(add-to-list 'auto-mode-alist '("\\.rs" . rust-ts-mode))
(add-hook 'rust-ts-mode-hook 'cargo-minor-mode)
(add-hook 'rust-ts-mode-hook 'eglot-ensure)
